### PR TITLE
Scan blocks with sapling keys and write the results to the database

### LINKED ISSUE #7953: Scan verified blocks from sapling activation to the tip
## Motivation

We want to scan all the blocks in the state, this is the ticket for scanning verified blocks from sapling activation to the tip.

### Suggested Design

Count from sapling activation height to `LatestChainTip.best_tip_height()`, ask the state for each block, then scan it. If there is no block for a height, wait 10 seconds, then try that height again in a loop.

Outside the state there is no distinction between finalised and non-finalised blocks, so this design will scan all of them.

## Out of Scope

- #7905

### LINKED ISSUE #8021: Write scanning results to database
After we have a database created in https://github.com/ZcashFoundation/zebra/issues/8019 and we have results from https://github.com/ZcashFoundation/zebra/issues/8018, we need to insert them to the database. 

This will be the issue that makes the connection between the scanning results and the persistent storage. Results can be written using the `Storage::add_sapling_result()` method.

Writing results to database files can block, so it should be wrapped in a `spawn_blocking()` function.

### LINKED ISSUE #8049: Check if any of the security issues are blockers for the MVP, and document the remaining security issues

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

This PR scans the blockchain and adds the scanner results to the scanner database.

Close #8021

I think it might also
close #7953
but @upbqdn feel free to replace that code with your PR, and re-open that ticket if it's not finished.

#### Check before marking the PR as ready for review:
  - [x] Will the PR name make sense to users?
  - [x] Does the PR have a priority label?
  - [x] Have you added or updated tests?
  - [x] Is the documentation up to date?

##### For significant changes:
  - [x] Is there a summary in the CHANGELOG?
  - [x] Can these changes be split into multiple PRs?

_If a checkbox isn't relevant to the PR, mark it as done._

The changelog summary for 1.5.0 already talks about scanning.

### Complex Code or Requirements

I tried scanning with all the keys to make sure it would work, but we only really need to scan with the full viewing key. I couldn't find the encoding for individual viewing keys, so I left that until later.

There's also a bunch of `spawn_blocking()` calls in here.

## Solution

Changes:
- Parse keys once at the start of the scan
  - Create a function that parses a key into a list of keys
  - Take a list of keys from the same key in the scanner for now
- Pass a ChainTipChange to the scanner function, this turned out to be mostly useless but it's good for progress
- Get a block from the state instead of the tip height
- Use a dummy sapling tree size
- Scan each block and add the results to storage
  - Convert a scanned block to a sapling result

Fixes:
- Fix availability of tokio::time in scanner
- Make it easier to pass keys and blocks to conversion functions
- Increase scanner wait times
- Don't log secret keys, only log every 100,000 blocks
- Move blocking tasks into spawn_blocking()

### Testing

- Update the acceptance test

I manually tested this, getting it in CI is ticket #8037

## Review

@oxarbitrage might want to review the scanner bits, and @upbqdn the block/state bits?

### Reviewer Checklist

Check before approving the PR:
  - [ ] Does the PR scope match the ticket?
  - [ ] Are there enough tests to make sure it works? Do the tests cover the PR motivation?
  - [ ] Are all the PR blockers dealt with?
        PR blockers can be dealt with in new tickets or PRs.

_And check the PR Author checklist is complete._

## Follow Up Work

#8022 then lots of testing:
- #7813 
- #8037 
- Round-trip tests for the database formats

### CODE REVIEW

**Comment by teor2345** [comment]
Seems to work in my manual tests, and quickly too:
> 2023-12-01T04:57:54.983238Z  INFO {zebrad="850bd6a" net="Main"}: zebra_scan::scan: Scanning the blockchain: now at block Height(600000), current tip Some((Height(2315746), block::Hash("00000000017abab20936b148ac1ed54a38cb87019f7ac489baeab2db256ea59e")))
> 2023-12-01T04:57:54.983251Z  INFO {zebrad="850bd6a" net="Main"}: zebra_scan::scan: Scanning the blockchain for key 0, started at block Height(419200)

This is only running on one core, and doing each key twice (dfvk and ivks).

File: zebra-scan/src/scan.rs [hunk 1]
```diff
@@ -1,81 +1,184 @@
...
+                    .best_tip_height_and_hash(),
             );
         }
 
-        tokio::time::sleep(CHECK_INTERVAL).await;
+        for (key_num, (sapling_key, birthday_height)) in key_birthdays.iter().enumerate() {
+            // # Security
+            //
+            // We can't log `sapling_key` here because it is a private viewing key. Anyone who reads
+            // the logs could use the key to view those transactions.
```
Comment (oxarbitrage) [inline zebra-scan/src/scan.rs:127]: Thanks for the fix. I initially added the pk to the log.

File: zebra-scan/src/scan.rs [hunk 2]
```diff
@@ -88,9 +191,9 @@ pub async fn start(mut state: State, storage: Storage) -> Result<(), Report> {
 /// - Add prior block metadata once we have access to Zebra's state.
 pub fn scan_block<K: ScanningKey>(
     network: Network,
-    block: Arc<Block>,
+    block: &Arc<Block>,
     sapling_tree_size: u32,
-    scanning_key: &K,
+    scanning_keys: &[K],
```
Comment (oxarbitrage) [inline zebra-scan/src/scan.rs:196]: I don't understand why it was better to reverse this change.

File: zebrad/tests/acceptance.rs [hunk 3]
```diff
@@ -2805,8 +2805,8 @@ async fn fully_synced_rpc_z_getsubtreesbyindex_snapshot_test() -> Result<()> {
 }
 
 /// Test that the scanner task gets started when the node starts.
-#[cfg(feature = "shielded-scan")]
 #[test]
+#[cfg(feature = "shielded-scan")]
```
Comment (oxarbitrage) [inline zebrad/tests/acceptance.rs:2809]: Some tests do this before the `[test]` macro (ex: https://github.com/ZcashFoundation/zebra/blob/main/zebrad/tests/acceptance.rs#L1305) and others do this after. What is the best way and why? Should we normalize the codebase? We should open a separated cleanup ticket for this if so.

File: zebrad/tests/acceptance.rs [hunk 4]
```diff
@@ -2832,13 +2832,16 @@ fn scan_task_starts() -> Result<()> {
     let output = child.wait_with_output()?;
 
     output.stdout_line_contains("spawning shielded scanner with configured viewing keys")?;
+    /*
+    TODO: these lines only happen when we reach sapling activation height
```
Comment (oxarbitrage) [inline zebrad/tests/acceptance.rs:2836]: optional:

I think we should print a message as soon as we know we are behind the sapling activation block saying something as:

"scan task will not get started until the blockchain is at sapling activation height (block number: {sapling activation_height})"

This way users will know the scanner will do nothing until that.

We can look for that message in this test so it will not require further modifications. A  new test will be done with state.

File: zebra-scan/Cargo.toml [hunk 5]
```diff
@@ -22,9 +22,10 @@ categories = ["cryptography::cryptocurrencies"]
 
 color-eyre = "0.6.2"
 indexmap = { version = "2.0.1", features = ["serde"] }
+itertools = "0.12.0"
 semver = "1.0.20"
 serde = { version = "1.0.193", features = ["serde_derive"] }
-tokio = "1.34.0"
+tokio = { version = "1.34.0", features = ["time"] }
```
Comment (oxarbitrage) [inline zebra-scan/Cargo.toml:28]: Is this change needed?

File: zebra-scan/src/scan.rs [hunk 6]
```diff
@@ -1,81 +1,184 @@
...
+            let mut storage = storage.clone();
+
+            // We use a dummy size of the Sapling note commitment tree.
+            //
+            // We can't set the size to zero, because the underlying scanning function would return
+            // `zcash_client_backeng::scanning::ScanError::TreeSizeUnknown`.
+            //
+            // And we can't set them close to 0, because the scanner subtracts the number of notes
+            // in the block, and panics with "attempt to subtract with overflow". The number of
+            // notes in a block must be less than this value, this is a consensus rule.
```
Comment (oxarbitrage) [inline zebra-scan/src/scan.rs:153]: Optional:

Quote the consensus rule. Use `# Consensus` doc tag as we use everywhere else.

File: zebra-scan/src/scan.rs [hunk 7]
```diff
@@ -1,81 +1,184 @@
...
+            let sapling_tree_size = 1 << 16;
+
+            tokio::task::spawn_blocking(move || {
+                let dfvk_res =
+                    scan_block(network, &block, sapling_tree_size, &dfvks).map_err(|e| eyre!(e))?;
+                let ivk_res =
+                    scan_block(network, &block, sapling_tree_size, &ivks).map_err(|e| eyre!(e))?;
```
Comment (oxarbitrage) [inline zebra-scan/src/scan.rs:160]: Ok, i see this is for testing purposes so it should be fine as it for now.

**Review by oxarbitrage — APPROVED** [review]
utACK. Looks really good to me, i left a few questions and optionals.

File: zebra-scan/src/scan.rs (hunk 2 above)
Comment (teor2345) [reply zebra-scan/src/scan.rs:196]: If we have a diversifiable full viewing key, there will only be one key. But if we have an individual viewing key, there will be two keys for the same account or transaction, the payment key and the change key (which are called external and internal).

So when I was testing using individual viewing keys, I needed to pass two of them into the one scan.

File: zebrad/tests/acceptance.rs (hunk 3 above)
Comment (teor2345) [reply zebrad/tests/acceptance.rs:2809]: It doesn't matter, it works the same way in either order, I just found it easier to read.

File: zebrad/tests/acceptance.rs (hunk 4 above)
Comment (teor2345) [reply zebrad/tests/acceptance.rs:2836]: Great, let's do this in a separate ticket

Comment (teor2345) [reply zebrad/tests/acceptance.rs:2836]: This is now #8041

File: zebra-scan/Cargo.toml (hunk 5 above)
Comment (teor2345) [reply zebra-scan/Cargo.toml:28]: Yes, if the crate is built individually with no features, it needs access to the tokio time APIs.

Sometimes this works anyway in CI, I'm not sure why. Possibly because we're building all targets.

File: zebra-scan/src/scan.rs (hunk 6 above)
Comment (teor2345) [reply zebra-scan/src/scan.rs:153]: Since this is a dummy value, I expect we'll replace it later with the actual value from the state. So I don't think it will go to the audit like this.

File: zebra-scan/Cargo.toml (hunk 5 above)
Comment (teor2345) [reply zebra-scan/Cargo.toml:28]: This is now tested by #8042, which might fail the first time due to this bug in main

File: zebra-scan/src/scan.rs (hunk 6 above)
Comment (teor2345) [reply zebra-scan/src/scan.rs:153]: Added TODO in 247684c492 in PR #8045

File: zebra-scan/src/scan.rs (hunk 2 above)
Comment (teor2345) [reply zebra-scan/src/scan.rs:196]: Documented in 329c59dd1a in PR #8045
